keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1316 @@
+
+## Proposal
+Most of these issues stem from missing a standardized way of detecting a start of the node drain.
+This KEP proposes the introduction of a NodeMaintenance object that would signal an intent to
+gracefully remove pods from given nodes. The application pods should then signal back that the pods
+are being removed or migrated from the node. The implementation should also utilize existing node's
+`.spec.unschedulable` field, which prevents new pods from being scheduled on such a node.
+
+We will focus primarily on `kubectl drain` as a consumer of the NodeMaintenance API, but it can
+also be used by other drain implementations (e.g. node autoscalers) or manually. We will first
+introduce the API and then later modify the behavior of the Kubernetes system to fix all the node
+drain issues we mentioned earlier.
+
+To support workload migration, a new controller should be introduced to observe the NodeMaintenance
+objects and then mark pods for migration or removal with a condition. The pods would be selected
+according to the node at first (`nodeSelector`), but the selection mechanism can be extended later.
+Controllers can then implement the migration. The advantage of this approach is that controllers do
+not have to be aware of the NodeMaintenance object (no RBAC changes required). They only have to
+observe pods they own and react by migrating them. The first candidate is a deployment controller,
+since its workloads support surging to another node, which is the safest way to migrate. This would
+help to eliminate downtime not only for single replica applications, but for HA applications as
+well.
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin I want to have a simple interface to initiate a node drain/maintenance without
+any required manual interventions. I want to have an ability to manually switch between the
+maintenance phases (Planning, Cordon, Drain, Drain Complete, Maintenance Complete). I also want to
+observe the node drain via the API and check on its progress. I also want to be able to discover
+workloads that are blocking the node drain.
+
+#### Story 2
+
+As an application owner, I want to run single replica applications without disruptions and have the
+ability to easily migrate the workload pods from one node to another.
+
+#### Story 3
+
+Cluster or node autoscalers that take on the role of `kubectl drain` want to signal the intent to
+drain a node using the same API and provide a similar experience to the CLI counterpart.
+
+#### Story 4
+
+I want to be able to use a similar approach for general descheduling of pods that happens outside
+of node maintenance.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
```

Comment:
- If someone deletes a NodeMaintenance whilst the drain is ongoing, do we remove the EvacuationRequested condition on Pods already marked?

Reply:
- If a Pod is marked `EvacuationInitiated` and the NodeMaintenance object(s) targeting that Pod are removed / postponed / whatever, what should happen?

(my opinion: **proceed**; evacuation has been initiated so we should continue it and expect a higher level abstraction to make a replacement Pod).

Reply:
Do you mean function-wise or name-wise?
- If the former, we need a way to go through the stages manually.
- If the later, I have tried to find a better name but could not decide on a better one. One of them is a phase, but they are meant as a cyclic event and I want the stage transitions to be fairly rigid and not to return back. So that there is a defined lifecycle of a single NodeMaintenance.

Reply:
I suppose in general yes. We do not know about the application behavior and it may not support cancellation (e.g. Pod is already terminating). Ultimately it should be up to the controller. It could react on the removal of the EvacuationRequest, but it doesn't have to..

Reply:
So I think it is better to always remove the `EvacuationRequest` condition, so that our intent is clear.

Reply:
added a mention for this